bundledatasrc_test.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"archive/zip"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/juju/testing"
+	gc "gopkg.in/check.v1"
+)
+
+type BundleDataSourceSuite struct {
+	testing.IsolationSuite
+}
+
+var (
+	_ = gc.Suite(&BundleDataSourceSuite{})
+
+	multidocBundle = `
+applications:
+  wordpress:
+    charm: wordpress
+  mysql:
+    charm: mysql
+    num_units: 1
+relations:
+  - ["wordpress:db", "mysql:server"]
+--- # overlay.yaml
+applications:
+  wordpress:
+    offers:
+      offer1:
+        endpoints:
+          - "some-endpoint"
+--- # overlay2.yaml
+applications:
+  wordpress:
+    offers:
+      offer1:
+        acl:
+          admin: "admin"
+          foo: "consume"
+`
+)
+
+func (s *BundleDataSourceSuite) TestReadMultiDocBundle(c *gc.C) {
+	specs := []struct {
```

Comment:
In my experience, table based tests like this aren't worth it. They make it too hard to just run one of the subtests, and they make it harder to look at the control flow. Better to have:
TestReadMultiDocBundleYaml(c *gc.C) {
  path := bundleDirPath(c, "wordpress-multidoc")
  source := LocalBundleDataSource(...)
  assertBundleSourceProcessed(c, source)
}
And then repeat that for
TestReadMultiDocBundleArchive
TestReadMultiDocBundleDirectory
TestReadMultiDocBundleStream

You get to still share all the same code via a helper function, and it makes the tests much easier to manage.